cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,271 @@
+param name string
+```
+
+Which provides template specific information on how the parameter is used. It also ensures that these templates behave nicely with other deployment tools (like creating a deployment from the portal or using `az` directly).
+
+We leverage the existing metadata system for bicep templates and parameters to allow some `azd` specific metadata to be applied to properties:
+
+```bicep
+@description('The Azure location to use for all resources.')
+@metadata({
+    azd: {
+        type: 'location'
+    }
+})
+param location string
+```
+
+The `type` property of the `azd` metadata object denotes that this value should be set to an Azure location name (e.g. "westus2") and is used by `azd` to offer a "pick a location from this list of all Azure locations). This composes with other validation decorators that you can apply, for example `@allowed()` which can be used to restrict the set of locations offered to the user to pick from (for cases where a template uses services that are not deployed to all regions):
+
+```bicep
+@description('The Azure location to use for all resources.')
+@allowed([
+    'westus2'
+    'centralus'
+])
+@metadata({
+    azd: {
+        type: 'location'
+    }
+})
+param location string
+```
+
+To start, we support the following types:
+
+`location`: The value should be the short name of an Azure location. All locations are supported by default, but the exiting `@allowed` may restrict this set.
+`azd-deployment-context`: The value should be an `azd` deployment context object. This is a JSON object that contains information about the project which may be useful for values when deploying. It will be described in further detail later.
+
+### Deployment Outputs
+
+When a deployment completes, the outputs from the deployment are stored in the `config.json` file under the "outputs" key (a sibling of "properties"). As we did on inputs, we re-use the schema of an outputs property of an ARM Deployment to save ourselves the trouble of inventing a new format.  When supporting Pulumi and Terraform we'll have to build an adapter but the format should be flexible enough to support this.
+
+### Environment Files
+
+An explicit design goal is that all environment state is stored in single JSON file, which should provide flexibility in developing some longer term story around collaboration.
+
+While no longer technically required, we continue produce a `.env` and `.parameters.json` in the `.azure\<environment-name>` folder use by external tooling. These are generated artifacts that are updated any time the authoritative values are updated in `config.json`.
+
+The generated `.env` file no longer contains deployment parameters.
+
+Now it is just the union of outputs from the deployment and any keys managed manually via `azd env set`. Since this file is regenerated each time a deployment happens, if a user removes an output from their `.bicep` file, the corresponding output will be removed from the `.env` file. 
```

Comment:
I assume the additional settings that are neither deployment parameters, nor deployment outputs, would go into its own section in the `config.json` file?